src/focus-change.cpp

```diff
@@ -0,0 +1,215 @@
+            const int32_t cy = bb.y + bb.height / 2;
+
+            const int32_t scan_w_intrm = scan_width.value() > 0 ?
+                scan_width.value() : scan_width.value() < 0 ?
+                cur_bb.width - abs(scan_width.value()) : cur_bb.width;
+            const int32_t scan_h_intrm = scan_height.value() > 0 ?
+                scan_height.value() : scan_height.value() < 0 ?
+                cur_bb.height - abs(scan_height.value()) : cur_bb.height;
+
+            const int32_t scan_w = std::max(scan_w_intrm / 2, 1);
+            const int32_t scan_h = std::max(scan_h_intrm / 2, 1);
+
+            const int32_t scan_w_l = cur_cx - scan_w;
+            const int32_t scan_w_h = cur_cx + scan_w;
+
+            const int32_t scan_h_l = cur_cy - scan_h;
+            const int32_t scan_h_h = cur_cy + scan_h;
+
+            const int32_t bias_up    = grace_up.value();
+            const int32_t bias_down  = grace_down.value();
+            const int32_t bias_right = grace_right.value();
+            const int32_t bias_left  = grace_left.value();
+
+            const bool w_cond = cx >= scan_w_l && cx <= scan_w_h;
+            const bool h_cond = cy >= scan_h_l && cy <= scan_h_h;
+
+            bool contains    = false;
+            int32_t distance = 0;
+            switch (orientation)
+            {
+              case orientation_t::UP:
+                contains = w_cond;
+                distance = cur_cy - cy - bias_up;
+                break;
+
+              case orientation_t::DOWN:
+                contains = w_cond;
+                distance = cy - cur_cy - bias_down;
+                break;
+
+              case orientation_t::RIGHT:
+                contains = h_cond;
+                distance = cx - cur_cx - bias_right;
+                break;
+
+              case orientation_t::LEFT:
+                contains = h_cond;
+                distance = cur_cx - cx - bias_left;
+                break;
+
+              default: /* unreachable */
+                ;
+            }
+
+            if ((distance >= 0) && contains)
+            {
+                if (distance < closest_cur)
+                {
+                    closest_cur = distance;
+                    new_focus   = view.get();
+                }
+            }
+        }
+
+        if (new_focus != nullptr)
+        {
+            wf::get_core().seat->focus_view(new_focus->self());
+        }
+    }
+
+    wf::key_callback on_key_up = [=] (auto)
+    {
+        change_focus(orientation_t::UP);
+        return true;
+    };
+
+    wf::key_callback on_key_down = [=] (auto)
+    {
+        change_focus(orientation_t::DOWN);
+        return true;
+    };
+
+    wf::key_callback on_key_right = [=] (auto)
+    {
+        change_focus(orientation_t::RIGHT);
+        return true;
+    };
+
+    wf::key_callback on_key_left = [=] (auto)
+    {
+        change_focus(orientation_t::LEFT);
+        return true;
+    };
+
+    void rebind()
```

Comment:
nitpick: I would probably drop the `re` prefix on this function name, since it's only called once.